### ISSUE TITLE
Signals

### ISSUE THREAD

**Issue by sondrelg** [samuelcolvin/arq#277]
Hi @samuelcolvin!

Would you be open to having signals added to arq? 

### My use case

I use [this request-id middleware](https://github.com/snok/asgi-correlation-id) to attach request ID values to logs, so each HTTP request's logs can be aggregated easily, making debugging a failed request much simpler. This works very well for all code executed by my web server, but there is one issue: passing work to a separate worker process breaks the implementation. Once work is sent to another process, we no longer have any reference to which request originally requested it, since the request ID is stored in-memory (which I think makes sense).

With Celery, this is actually pretty simple to fix: you can used their signals (and their `headers` object) to pass state from your issuing process to individual workers (attached to the task object), like this:

```python
from typing import Any, Dict
from uuid import uuid4

from celery import Task
from celery.signals import before_task_publish, task_prerun, task_postrun

from asgi_correlation_id.context_vars import correlation_id

header_key = 'CORRELATION_ID'


@before_task_publish.connect
def transfer_correlation_id(headers: Dict[str, str], **kwargs: Any) -> None:
    """
    Attach request ID to the task headers.
    """
    cid = correlation_id.get() 
    if cid:
        headers[header_key] = cid


@task_prerun.connect
def load_correlation_id(task: Task, **kwargs: Any) -> None:
    """
    Set request ID from the task header, if it exists.
    """
    id_value = task.request.get(header_key)
    if id_value:
        correlation_id.set(id_value)
    else:
        generated_correlation_id = uuid4().hex
        correlation_id.set(generated_correlation_id)


@task_postrun.connect
def cleanup(**kwargs: Any) -> None:
    """
    Clear context vars, to avoid re-using values in the next task.
    """
    correlation_id.set(None)
```

This seems like it would be simple to add to ARQ as well. Would you accept a PR for something like this?

**Comment by sondrelg** [comment]
I noticed after typing this out, that https://github.com/samuelcolvin/arq/pull/274 is somewhat related here, though it won't resolve my issue.

**Comment by sondrelg** [comment]
If accepted, I have two follow up questions:

- To solve my use case I need a `before_task_publish`, `task_prerun`, and `task_postrun` signals, plus I would need to be able to pass state from the enqueueing process to the worker; but adding 1 or 10 signals would basically be the same work, so perhaps it would make sense to build a complete suite of hooks right away? If so, do you have suggestions or ideas wrt. what events to include?

- In terms of the implementation, would you prefer that these were added as `WorkerSettings` attributes, mirroring the `on_startup` and `on_shutdown` hooks?

**Comment by sondrelg** [comment]
Quick update: after looking over the arq internals a little bit, it looks like the `enqueue_job` kwargs is sufficient for state transfer. Adding a `pre_publish` event hook ~[here](https://github.com/samuelcolvin/arq/blob/master/arq/connections.py#L151), where one could pass in and return kwargs would solve my use-case, assuming #274 was merged in its current state.

**Comment by sondrelg** [comment]
Since #274 was merged, all this would require is:

1. For a "on-enqueue"-style hook to be added
2. For there to be a way to set state in that hook, that persists to the worker when running `task_prerun`

**Comment by JonasKs** [comment]
We're also very interested in this. Happy to join you on a PR for this, @sondrelg if the idea is accepted by @samuelcolvin. Should probably wait until #259 is finalized/aioredis 2 support is added though.

**Comment by samuelcolvin** [comment]
Humm, thinking about this, I'd rather it was structured like middleware in http frameworks.

I'm intending to work on total rewrite of arq to be very close to fastapi but for long running tasks, I'll include middleware in that.

I don't think there' much point in working on a PR now.

**Comment by JonasKs** [comment]
That sounds very interesting, and is what we want too! Let us know if we can do anything.

**Comment by sondrelg** [comment]
Just out of curiosity, how will a middleware concept work for something like event hooks, where you need things to happen in distinct places in an events lifecycle?

Also, this isn't incredibly urgent for me, but do you have rough timeline in mind for when you'll do the rewrite?

**Comment by samuelcolvin** [comment]
Well like middleware, you could do stuff before or after the job or catch errors and do stuff with them.

Timeframe is hard to say, hopefully in the next month or two.

**Comment by teror4uks** [comment]
If I understood you correctly you could propagate a variable to a job by using `ctx` dict and then set [contextvar ](https://docs.python.org/3.8/library/contextvars.html) , also write a custom LogHandler to receive correlation id from the contextvar.

**Comment by sondrelg** [comment]
Yes that's right; but a package cannot make this the default behavior for all tasks queued with the `enqueue_job` method call, which is what I'd like to enable. 

In other words, I would like there to be a way for the `asgi-correlation-id` package to automatically pass this information through to the job and load it into the correct contextvar, without the user needing to do anything extra.

**Comment by waza-ari** [comment]
I'm in the exact same situation, FastAPI, want to use the library mentioned above to get some correlation id and want to pass this to arq jobs. I understand that there is currently no PR being worked on as the re-write should solve the issue, but did you guys found a workaround to make this work in the meantime?

**Comment by waza-ari** [comment]
Thank you @JonasKs , that was extremely helpful. I basically was able to nearly use exactly what you've posted. I'll post my complete solution here in case anyone else is looking for it

Worker Tasks

```python
from uuid import uuid4

from asgi_correlation_id import correlation_id

async def job(cid: uuid4):
    correlation_id.set(cid)
    # business logic ..
```

FastAPI:

```
from asgi_correlation_id import correlation_id

async def my_view():
    arq_redis.enqueue_job('job', cid=correlation_id.get())
```

In addition to that, I was also using arq cronjobs, which might call additional jobs as well. I wanted to have a correlation ID as well for them, which was a bit more complex:

```python
from functools import wraps

def partial(f, *args, **kwargs):
    """
    Wrapper for cron functions to inject corellation_id
    """

    @wraps(f)
    async def partial_function(ctx):
        return await f(*args, **kwargs)

    return partial_function

class WorkerSettings:  # pylint: disable=too-few-public-methods
    """
    Settings for the ARQ worker.
    """

    cron_jobs = [
        cron(partial(your_job, {}, cid=uuid4().hex), hour={1}, minute={12}),
        ...
   ]

**Comment by JonasKs** [comment]
For corn jobs you could also do this:

```py

async def set_correlation_id(ctx: 'Ctx') -> None:
    correlation_id.set(uuid4().hex)


class WorkerSettings:
    on_job_start = set_correlation_id

```

**Comment by waza-ari** [comment]
I probably could, but some of the jobs can be called both as cron and manually from FastAPI, that's why I thought using the `on_job_start` hook would not work, as then the job could not be called from FastAPI anymore
